File: app/src/main/java/com/lumiyaviewer/lumiya/slproto/messages/SimulatorSetMap.java
package com.lumiyaviewer.lumiya.slproto.messages;

import com.lumiyaviewer.lumiya.slproto.SLMessage;
import java.nio.ByteBuffer;
import java.util.UUID;

/**
 * SimulatorSetMap
 * simulator -> dataserver
 * reliable
 * Used to upload a map image into the database (currently used only for Land For Sale)
 *
 * <p>Template: {@code SimulatorSetMap Low 6 Trusted Unencoded}
 * (recovered/reference/message_template.msg).
 */
public class SimulatorSetMap extends SLMessage {
    public MapData MapData_Field;

    /** Block MapData, Single. */
    public static class MapData {
        public UUID MapImage; // LLUUID
        public long RegionHandle; // U64
        public int Type; // S32
    }

    public SimulatorSetMap() {
        this.zeroCoded = false;
        this.MapData_Field = new MapData();
    }

    @Override
    public int CalcPayloadSize() {
        return 32;
    }

    @Override
    public void Handle(SLMessageHandler sLMessageHandler) {
        sLMessageHandler.HandleSimulatorSetMap(this);
    }

    @Override
    public void PackPayload(ByteBuffer byteBuffer) {
        // Message number: Low 6 (SimulatorSetMap).
        byteBuffer.putShort((short) 0xFFFF);
        byteBuffer.put((byte) 0x00);
        byteBuffer.put((byte) 0x06);
        packLong(byteBuffer, this.MapData_Field.RegionHandle);
        packInt(byteBuffer, this.MapData_Field.Type);
        packUUID(byteBuffer, this.MapData_Field.MapImage);
    }

    @Override
    public void UnpackPayload(ByteBuffer byteBuffer) {
        this.MapData_Field.RegionHandle = unpackLong(byteBuffer);
        this.MapData_Field.Type = unpackInt(byteBuffer);
        this.MapData_Field.MapImage = unpackUUID(byteBuffer);
    }
}

File: app/src/main/java/com/lumiyaviewer/lumiya/slproto/messages/SimulatorSetMap.kt
package com.lumiyaviewer.lumiya.slproto.messages

import com.lumiyaviewer.lumiya.slproto.SLMessage
import java.nio.ByteBuffer
import java.util.UUID

/**
 * SimulatorSetMap
 * simulator -> dataserver
 * reliable
 * Used to upload a map image into the database (currently used only for Land For Sale)
 *
 * <p>Template: {@code SimulatorSetMap Low 6 Trusted Unencoded}
 * (recovered/reference/message_template.msg).
 */
open class SimulatorSetMap : SLMessage() {
    @JvmField var MapData_Field: MapData = MapData()

    /** Block MapData, Single. */
    open class MapData {
        @JvmField var MapImage: UUID? = null
        @JvmField var RegionHandle: Long = 0L
        @JvmField var Type: Int = 0
    }

    init {
        zeroCoded = false
    }

    override fun CalcPayloadSize(): Int {
        return 32
    }

    override fun Handle(messageHandler: SLMessageHandler) {
        messageHandler.HandleSimulatorSetMap(this)
    }

    override fun PackPayload(byteBuffer: ByteBuffer) {
        // Message number: Low 6 (SimulatorSetMap).
        byteBuffer.putShort(0xFFFF.toShort())
        byteBuffer.put((0x00).toByte())
        byteBuffer.put((0x06).toByte())
        packLong(byteBuffer, MapData_Field.RegionHandle)
        packInt(byteBuffer, MapData_Field.Type)
        packUUID(byteBuffer, MapData_Field.MapImage)
    }

    override fun UnpackPayload(byteBuffer: ByteBuffer) {
        MapData_Field.RegionHandle = unpackLong(byteBuffer)
        MapData_Field.Type = unpackInt(byteBuffer)
        MapData_Field.MapImage = unpackUUID(byteBuffer)
    }
}
